### ISSUE TITLE
xRegistry - writing binary data with '00' seems to have strange effects

### ISSUE THREAD

**Issue by marcinbojko** [PowerShell/xPSDesiredStateConfiguration#301]
When we try to write data using xRegistry like in example:

```yaml
tightvncpassword:
  dsc_key: HKEY_LOCAL_MACHINE\SOFTWARE\TightVNC\Server
  dsc_valuename: Password
  dsc_hex: true
  dsc_valuedata: 00 00 00 00 00 00 00 00
  dsc_valuetype: binary
```

Setting registry values has different effect if we are going to use '00' or not.
Let's say we want to set value to: 'aabbccdd'

We'll have:

> VERBOSE: [WIN-LO53EUODSPV]:                            [[xRegistry]DirectResourceAccess] (SET) Set registry key value 'HKLM:\SOFTWARE\TightVNC\Server\ControlPassword' to 
> '(170, 187, 204, 221)' of type 'binary'

Let's change our value to: '00bbccdd'

> VERBOSE: [WIN-LO53EUODSPV]:                            [[xRegistry]DirectResourceAccess] (SET) Set registry key value 'HKLM:\SOFTWARE\TightVNC\Server\ControlPassword' to 
> '(187, 204, 221)' of type 'binary'

First '00' get's trimmed to ''.
Single '0' instead of '00' ?
Same effect.
What about line of zeroes like: '00000000'?

> VERBOSE: [WIN-LO53EUODSPV]:                            [[xRegistry]DirectResourceAccess] (SET) Set registry key value 'HKLM:\SOFTWARE\TightVNC\Server\ControlPassword' to 
> '()' of type 'binary'

Value: 'aabbcc00dd000000' - seems to be legit.

> VERBOSE: [WIN-LO53EUODSPV]:                            [[xRegistry]DirectResourceAccess] (SET) Set registry key value 'HKLM:\SOFTWARE\TightVNC\Server\ControlPassword' to 
> '(170, 187, 204, 0, 221, 0, 0, 0)' of type 'binary'
> 

Looks like we have problem if the first byte in binary is '00' or '0'

**Comment by Amedama96** [comment]
Hi @marcinbojko 
I tested all patterns you mentioned.
However, the issue did not reproduce.

please try the following
- If you are using older version of xRegistry, update to 5.1.0.0 or later.
- If you are using PS with another CM tools like Puppet, Try to use pure Powershell only.

This is the PS script I used  for testing
```Powershell
Configuration TestConfig {
    Import-DscResource -ModuleName xPsDesiredStateConfiguration
    Node localhost {
        xRegistry TestBinReg
        {
            Ensure = 'Present'
            Key = 'HKEY_LOCAL_MACHINE\SYSTEM\Test'
            ValueName = 'Test'
            ValueData = '00bbccdd'
            ValueType = 'Binary'
            Hex = $true
        }
    }
}
TestConfig -OutputPath 'C:\DSC'
Start-DscConfiguration -Path 'C:\DSC' -Verbose -Wait -Force
```

**Comment by marcinbojko** [comment]
Hi @Amedama96 
I have to use version included in https://github.com/puppetlabs/puppetlabs-dsc, so - can we wait a little longer for their new release? If the issue will not be reproductible in 5.1.0.0 (with mentioned module) - case closed.

**Comment by Amedama96** [comment]
I think the issue related in #276 and already fixed in #299

**Comment by marcinbojko** [comment]
I think you're right. Should we proceed with closing this one, as it is already fixed? Using v.5.0 is just an artefact from puppetlabs-dsc.

**Comment by kwirkykat** [comment]
@marcinbojko Not sure puppetlabs-dsc has updated to either of the more recent versions yet (looks like they are still on 5.0.0.0). But yes this issue should be fixed in 5.1.0.0+. We have also added tests for the binary 0 case now to make sure this functionality doesn't regress. You ok with closing this issue?

**Comment by marcinbojko** [comment]
Please, do.  
                
              
               
                 
                  Dnia 16 lutego 2017 19:59 Katie Keim &lt;notifications@github.com&gt; napisał(a):
                 
                 
                   github.com @marcinbojko  Not sure puppetlabs-dsc has updated to either of the more recent versions yet (looks like they are still on 5.0.0.0). But yes this issue should be fixed in 5.1.0.0+. We have also added tests for the binary 0 case now to make sure this functionality doesn&#39;t regress. You ok with closing this issue?
